### ISSUE TITLE
Opensnitch UI is freezing when internet connection is blocked (e.g. by Mullvad VPN)

### ISSUE THREAD

**Issue by Shadow505** [evilsocket/opensnitch#565]
Hi,

I have been having a very annoying issue for a while now and just found out what is causing it.

The issue is that the Opensnitch UI is freezing whenever I connected to a different mullvad server that I have not yet whitelisted in opensnitch.

I just found out what this is causing this: When connecting to a new server, mullvad will temporarily block the internet connection. If the target IP (mullvad server) has already been whitelisted this won't be an issue because opensnitch won't have to present the dialog that allows you to allow/deny the connection.

However, if the IP has not been whitelisted yet, opensnitch has to open the dialog, so that you can choose to allow/deny the connection attempt. The issue is that, while Mullvad is blocking the internet connection, the Opensnitch UI is completely frozen. Sometimes you get lucky and the dialog still pops up but you won't be able to select anything. After the timeout has been reached that has been specified in opensnitch, it will automatically allow/disallow the connection (depending on your opensnitch settings). This is obviously an issue.

If you open the Mullvad settings and enable the "VPN always required" option and then go ahead and then click on Disconnect, Mullvad will block your internet connection (just as explained above). Now the Opensnitch UI is again completely frozen and won't even recover if you connect to a Mullvad server again to unblock the internet connection. You pretty much have to kill the opensnitch-ui process and start it again.

There are also entries like this in the log:

`ERR  Breaking because endless loop was detected in deleteOld`

This issue happens on the 1.5.0 and the ones before that, e.g. 1.4.3. I also installed the ebpf kernel module and have enabled it in the settings.

My OS is Arch Linux. Currently running on kernel 5.15.7.

Best regards

**Comment by gustavo-iniguez-goya** [comment]
> If you open the Mullvad settings and enable the "VPN always required" option and then go ahead and then click on Disconnect, Mullvad will block your internet connection (just as explained above). Now the Opensnitch UI is again completely frozen and won't even recover if you connect to a Mullvad server again to unblock the internet connection. You pretty much have to kill the opensnitch-ui process and start it again.

I've tried to reproduce this problem, but opensnitch GUI doesn't freeze in my case.

What happens if you disable "VPN always required" option?
Do you see "Kernel connection" pop-ups when trying to connect to the VPN? https://github.com/evilsocket/opensnitch/issues/454#issuecomment-993516122
Does it work if you close and reopen the GUI?

One reason for such blockages is if there're too much events recorded, specially if the events are saved to disk.

**Comment by gustavo-iniguez-goya** [comment]
>There are also entries like this in the log:
>ERR Breaking because endless loop was detected in deleteOld

It shouldn't be related, but anyway, we're working to fix that #550

**Comment by Shadow505** [comment]
> > If you open the Mullvad settings and enable the "VPN always required" option and then go ahead and then click on Disconnect, Mullvad will block your internet connection (just as explained above). Now the Opensnitch UI is again completely frozen and won't even recover if you connect to a Mullvad server again to unblock the internet connection. You pretty much have to kill the opensnitch-ui process and start it again.
> 
> I've tried to reproduce this problem, but opensnitch GUI doesn't freeze in my case.
> 
> What happens if you disable "VPN always required" option? Do you see "Kernel connection" pop-ups when trying to connect to the VPN? [#454 (comment)](https://github.com/evilsocket/opensnitch/issues/454#issuecomment-993516122) Does it work if you close and reopen the GUI?
> 
> One reason for such blockages is if there're too much events recorded, specially if the events are saved to disk.

Hi,

I do see the "Kernel connection" popup but I'm not able to properly make any choice due to the lag/freeze.

And regarding: 
> One reason for such blockages is if there're too much events recorded, specially if the events are saved to disk.

Not sure what you consider "too many events", too me that is definitely a bug anyway. In my opinion, the UI should not become unusable/unresponsive under any circumstance.


The UI is generally extremely laggy, to the point of becoming almost unusable for certain tasks. For example, I just accidentally blocked port 443 (without knowing that I did) and while trying to find and delete the corresponding rule, the UI is constantly lagging, for example when scrolling through the rules. 

There is also a very annoying issue that causes the UI to jump between scroll positions, that makes it extremely difficult to work with the data, e.g. to find something, delete it, etc. I have recorded a video so that you can see it. Just so you know, I'm not scrolling by myself unless you actually see me drag the scrollbar.

https://user-images.githubusercontent.com/9123343/146455179-60d098b6-afad-4682-b2e0-763bac366801.mp4

The best solution would be obviously if this scrolling can be prevented without the user having to take any action. I'm guessing it occurs because the list is frequently refreshed, so if this cannot be properly fixed, a workaround might be a button to temporarily disable updating of the table.

Thank you for your time!

Best regards

**Comment by gustavo-iniguez-goya** [comment]
Hi, 

> Not sure what you consider "too many events", too me that is definitely a bug anyway. In my opinion, the UI should not become unusable/unresponsive under any circumstance.

I agree. On the Events tab there's a box to limit how many events are displayed, but that combo is not present on the other tabs, so if you have thousands (>50k) of entries on the Hosts/Ports/etc tab, the GUI will freeze. I'm working on a fix for this.problem.

Thank you for the video! it's very clarifying. That's caused by displaying more than 256 rows. I'll also work on a fix for that problem next week.

**Comment by gustavo-iniguez-goya** [comment]
> ERR Breaking because endless loop was detected in deleteOld

fixed here 7c87baff13f8743597f39649c136e6f7f02f7223

> There is also a very annoying issue that causes the UI to jump between scroll positions, that makes it extremely difficult to work with the data,

fixed here deee72a

> Not sure what you consider "too many events", too me that is definitely a bug anyway

With deee72a changes , now the GUI is much more usable and useful. But still far from perfect.

Now I can list queries with >10k results without the GUI getting frozen, but it still lags quite a bit. I'll try to improve it.

Regarding selection of rows and auto-deselecting on refreshing the view: as far as I can tell, it depends on the PyQt version of your system. With `5.15.6` now when you select a row it's not deselected when the a view is refreshed. But with 5.5.1 and 5.13 rows are deselected after refreshing.

**Comment by gustavo-iniguez-goya** [comment]
@Shadow505 , could you test latest changes and let me know if they have improved the problems you had? 

Could you also tell me if you were using an in-memory DB or if you were saving events to a DB on disk?

**Comment by Shadow505** [comment]
Hi,

I just update to the latest commit (#434bbf9) and unfortunately I am no longer able to open the UI. The error is the following:

```
Traceback (most recent call last):
File "/usr/bin/opensnitch-ui", line 22, in <module>
from opensnitch.service import UIService
File "/usr/lib/python3.10/site-packages/opensnitch/service.py", line 17, in <module>
from opensnitch.dialogs.stats import StatsDialog
File "/usr/lib/python3.10/site-packages/opensnitch/dialogs/stats.py", line 18, in <module>
from opensnitch.customwidgets.main import ColorizedDelegate, ConnectionsTableModel
ModuleNotFoundError: No module named 'opensnitch.customwidgets'
```

So.. Yeah.. 😭

**Comment by gustavo-iniguez-goya** [comment]
Merry Christmas to you too! 

Check if you have `/usr/lib/python3/dist-packages/opensnitch/customwidgets.py` . If it exists delete it.

**Comment by Shadow505** [comment]
Thank you! :)

I do not appear to have a dist-packages folder in any of the /usr/lib/python3X directories.

I did notice something a couple of minutes though, while trying to find the cause of the error. The package is installed in the `/usr/lib/python3.10/site-packages/opensnitch/` directory. Is that directory supposed to contain a _customwidgets_ directory? Because there isn't one. If it is supposed to be there, then the PKGBUILD from the AUR might be outdated?

If it's neither of the above, then 🤷

**Comment by gustavo-iniguez-goya** [comment]
> Is that directory supposed to contain a customwidgets directory? 

Yes, it should exist, maybe it's outdated. You can clone the repo and copy the missing files:
```
$ cd /tmp
$ git clone https://github.com/evilsocket/opensnitch
$ sudo cp -r opensnitch/ui/opensnitch/customwidgets/ /usr/lib/python3.10/site-packages/opensnitch/
```

**Comment by Shadow505** [comment]
Yep, that was my plan also, just wanted to make sure :D

Maybe you could take a look at the PKGBUILD in the AUR to see what's wrong about it, so that it can be fixed. I was assuming that the package in the AUR is maintained by you (or your colleagues/friends).

Please see https://github.com/evilsocket/opensnitch/issues/566#issuecomment-1001281692, I just added additional info to that.

I will let you know later/tomorrow if the other issues (e.g. the ones mentioned in this github issue) have been solved/improved.

Best regards :)

**Comment by Shadow505** [comment]
I decided to stay up a little longer in hopes to provide useful info before going to sleep. Unfortunately the issue with the freezing UI has not been fixed. The UI still locks up as soon as I connect to a new Mullvad server (without an existing rule).

![Screenshot_298](https://user-images.githubusercontent.com/9123343/147428887-55637dcc-218c-42d6-83c5-e1ddb3f474c3.png)

**Comment by gustavo-iniguez-goya** [comment]
ok, problem found @Shadow505 , I forgot to add a file 0:) 

I've updated opensnitch-git package and now it works fine. It should work for you too, so just delete the package and install it again. Thank you for reporting this!

On the other hand, I still don't reproduce that blockages. Do you use in-memory DB or do you save events to a DB on disk?

**Comment by Shadow505** [comment]
@gustavo-iniguez-goya,

so.. I just spent about 3 hours trying different things out, while also trying to get a useful screen recording.

In the end I did the following: I removed the existing database and also removed the rules from /etc/opensnitchd/rules. I closed pretty much any process that might actively using the connection to avoid that the UI is being filled with events that will make it harder to understand what's going on when looking at the recorded footage. For this recording I also changed the storage type to Memory.

The result appears to be the following: When I connect to a new server (for which there are no rules yet), it takes a bit until the Opensnitch popup appears. The UI will stay responsive but there appears to be a rather nasty bug: While the countdown is running (to make a choice), for some reason Opensnitch decides to allow the connection, even though I did not make any decision yet. You can see that Mullvad is connected after a couple of seconds, while the Popup is still there with plenty time left. 

Neither the UI, nor the log appear to be logging anything related to that. I repeated this a couple of times, always with the same result.

Also you might notice that there is another bug in the UI which prevents it from showing new events until I manually do something that appears to be triggering a refresh, for example setting the amount of records to 100 and immediately back to 50. At 0:55 can can also notice another bug in the UI: As soon as I click the deny button, the currently displayed row in the UI (which has "Allow" as the action) is "overwritten" with the rule corresponding to the "Deny" action that I performed. Only after performing the above workaround, both of them appear. Also at 00:42, take a look at the bottom left of the UI where it states the amount of connections. It jumps from 2 connections to 12 connections as soon as Opensnitch decides to allow the connection, even without me wanting to do so. 

If I find out what could be causing the freezing of the UI, I will let you know. It did not happen in this case, which is not really that useful because that cannot be considered as a productive setup due to all the measures I took. But the issue that opensnitch just let's the connection go through seems rather critical.

I do really hope that this information helps you, due to the fact that it took multiple hours to get it.

https://user-images.githubusercontent.com/9123343/147514058-18d81087-4207-4677-aed4-f7d33267f47c.mp4

Best regards

**Comment by Shadow505** [comment]
Additionally see #579 for another UI bug.

**Comment by gustavo-iniguez-goya** [comment]
Thank you for the video! it's really very clarifying, it helps a lot to understand a little bit better the problem.

>  for some reason Opensnitch decides to allow the connection, even though I did not make any decision yet. You can see that Mullvad is connected after a couple of seconds, while the Popup is still there with plenty time left.

What I think is happening here is that as you have selected "Allow" as default Action in the Preferences, opensnitch is allowing the connections of the VPN before closing the pop-up.

The logic is as follow: 
- if the daemon is connected to the GUI -> apply the default action while the pop-up is active, of if the timeout fires.
- if the GUI is not active, apply daemon's default action.

If you change default actions to Deny, both daemon's and GUI's, it shouldn't allow connections until the pop-up is closed.

Regarding the Events view not being updated, I think you don't have latest sources. That problem was fixed a couple of commits ago. You can clone the repo, and compare `opensnitch/ui/opensnitch/` files with the ones you have in `/usr/lib/python3.10/site-packages/opensnitch/` .

**Comment by Shadow505** [comment]
Hi,

`if the daemon is connected to the GUI -> apply the default action while the pop-up is active`

I'm sorry but that does not make much sense to me... Why would it apply the default action while the popup is still open and plenty of time left to make a choice? It should not apply the default action unless either timeout is exceeded or the UI has not been started. If it automatically and immediately applies the default action while showing me the popup then there is no reason to show me the popup at all. The point of the popup is to let the user decide how to handle this specific request. This is clearly a bug.

`I think you don't have latest sources. That problem was fixed a couple of commits ago`

Yep, you have pushed a couple of commits while I was already working on this. At that time the above errors where produced, I will try again later if the new commits fix this.

**Comment by gustavo-iniguez-goya** [comment]
> Why would it apply the default action while the popup is still open and plenty of time left to make a choice? 

It's a requirement. Whenever we intercept a connection, the system expects an answer on what to do with it, allow or deny it.
While the pop-up is active, the application may open more connections, so we need to answer them with an action.

The only way of stopping it completely would be sending a SIGSTOP signal to the process, freezing completely the application until the user emits a verdict.

If you want to block the connections of an app while the pop-up is active, configure default action to Deny.

In the end, if you set default action to Allow I guess you just want to be warned about new outgoing connections, create the rule for you if you don't answer the pop-up and decide if Deny it if you decide so.

IIRC, this was already discussed and changed, according to other people's preferences.

**Comment by gustavo-iniguez-goya** [comment]
> I decided to stay up a little longer in hopes to provide useful info before going to sleep. Unfortunately the issue with the freezing UI has not been fixed. The UI still locks up as soon as I connect to a new Mullvad server (without an existing rule).
> 
> Also, the UI in general is still very unresponsive, for example if you scroll through your rules or pretty much do anything else in the UI :(

I can confirm that this freezing problems occur when using the DB from disk. They're caused mainly by this function when it have to process a large amount of events, where "large amount" >= ~10k records (tottime column in seconds):
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       6 1156.374  192.729 1185.405  197.568 /home/ga/Proiektuak/opensnitch/gui/opensnitch/ui/lib/python3.9/site-packages/opensnitch/customwidgets/main.py:246(updateDistinctIfNeeded)
```

https://github.com/evilsocket/opensnitch/blob/c1d6d943e8eacf6d41b5537693eba48027ccfda7/ui/opensnitch/customwidgets/main.py#L245

Removing it from the equation, the next bottleneck (specially with "big" DB files, > 200/300MB) is the function to check if the DB is corrupted (tottime column in seconds):
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     2   25.394   12.697   48.861   24.430 /home/ga/Proiektuak/opensnitch/gui/opensnitch/ui/lib/python3.9/site-packages/opensnitch/database.py:47(is_db_ok)
```

So due to this, everything slowdowns. All the problems you described are caused by this, so for now don't save events to disk.
I've added an option to limit the number of events to save to disk. But even so, depending on your use case, it might not be enough.

I'll try to improve the performance. Any help very much welcome!

(Note: I'm not justifying what is "big" or "large amount", I'm only describing what's going on).

**Comment by gustavo-iniguez-goya** [comment]
I believe that this issue has been resolved in v1.5.0rc2. If you still experience similar problems, please, open a new issue.
